util/src/main/scala/scala/scalanative/util/Scope.scala

```diff
@@ -39,21 +41,34 @@ object Scope {
       throw new UnsupportedOperationException("Can't close forever Scope.")
   }
 
+  /** Unsafe manually managed scope.*/
+  def unsafe: Scope = new Impl {}
+
   private sealed class Impl extends Scope {
-    private[this] var resources: List[Resource] = Nil
+    type Resources = List[Resource]
+
+    private[this] val resources = new AtomicReference[Resources](Nil)
+
+    private def set(fn: Resources => Resources): Resources =
+      resources.getAndUpdate {
+        new UnaryOperator[Resources] {
+          override def apply(t: Resources): Resources = fn(t)
+        }
+      }
 
     def acquire(res: Resource): Unit = {
-      resources ::= res
+      set(res :: _)
     }
 
-    def close(): Unit = resources match {
+    def close(): Unit = resources.get() match {
       case Nil =>
         ()
 
-      case first :: rest =>
-        resources = rest
+      case _ =>
+        val first :: _ = set(_.tail)
         try first.close()
         finally close()
     }
```

Comment:
This could be more thread-safe by taking a snapshot of `resources` only once:
```scala
def close(): Unit = {
  def loop(resources: Resources): Unit = resources match {
    case Nil =>
      ()
    case first :: rest =>
      try first.close()
      finally loop(rest)
  }
  loop(resources.getAndSet(Nil))
}